04. Deviseによるユーザー認証プラクティス

note103 opened the pull request:
## Todo

- [x] メールアドレスとパスワードを入力してアカウント登録ができる
  - [x] アカウント登録時に郵便番号、住所と自己紹介文を登録できるようにしても良い
- [x] メールアドレスとパスワードでログインできる
- [x] ログインしたユーザーはログアウトできる
- [x] ユーザーの一覧画面と詳細画面を作成する
  - [x] 一覧ページはページング処理を付ける
- [x] アプリの利用にはログインを必須とする
  - ログインせずに本の一覧画面やユーザーの一覧画面を開くことはできない
- [x] アカウント編集ページで自分のメールアドレスとパスワード、郵便番号、住所と自己紹介文を編集できる
  - 自分以外のユーザーの情報は編集できない
- [x] パスワードを忘れたらパスワード再設定メールを経由してパスワードを再設定できる
- [x] Deviseが用意した画面も日本語で表示する。それ以外の新しく追加した画面もi18nで日本語化する
- [x] いくつかの画面のスクリーンショットを載せる
  - どの画面を載せるかは任意
- [x] rubocopとerblintをパスさせる
  - 要スクリーンショット

## Links

### ストロングパラメーター
- [ユーザー管理機能のためのdevise導入手順](https://zenn.dev/furum/articles/8e6fc52225422b#%E3%82%B9%E3%83%88%E3%83%AD%E3%83%B3%E3%82%B0%E3%83%91%E3%83%A9%E3%83%A1%E3%83%BC%E3%82%BF%E3%83%BC%E3%81%AE%E8%A8%AD%E5%AE%9A%EF%BC%88%E3%82%AB%E3%83%A9%E3%83%A0%E3%82%92%E8%BF%BD%E5%8A%A0%E3%81%97%E3%81%9F%E5%A0%B4%E5%90%88%EF%BC%89)

### ユーザー編集後のリダイレクト
- [Deviseをカスタマイズして、アカウントを編集した後のリダイレクト先を変更する@TECH CAMP #24｜とだ](https://note.com/ruquia7/n/n4838547cb054)
- [heartcombo/devise: Flexible authentication solution for Rails with Warden.](https://github.com/heartcombo/devise#strong-parameters)

### Fakerによるプロフィール例文生成
- [faker/doc/japanese_media/studio_ghibli.md at main · faker-ruby/faker](https://github.com/faker-ruby/faker/blob/main/doc/japanese_media/studio_ghibli.md)

## How to use

- [README](https://github.com/note103/fjord-books_app-2023/blob/main/README.md)にあるとおりですが、ユーザー一覧のページング処理を確認するため、`seeds.rb` にユーザー生成用のコードを追加しています。
- 任意のタイミングで `rails db:reset` または `rails runner db/seeds.rb` をしていただければ `seeds.rb` 内のダミーデータがアプリに読み込まれます。（書籍データ50件・ユーザーデータ30件）

note103 committed:
devise、devise-i18n gemを追加

note103 committed:
Deviseをインストール

note103 committed:
action_mailerのデフォルトURLオプションを設定

note103 committed:
ルートパスを"books#index"に設定

note103 committed:
Userモデルを追加

note103 committed:
db:migrateを実施

note103 committed:
before_actionでユーザー認証を追加

note103 committed:
ユーザーがサインインしている場合のナビゲーションバーを追加

note103 committed:
ユーザーコントローラーとビューを追加

note103 committed:
ユーザーリソースのルーティングを追加

note103 committed:
ナビゲーションバーにマイページとユーザー一覧、書籍一覧へのリンクを追加

note103 committed:
ユーザーモデルの翻訳を追加

I18n::InvalidPluralizationData対策

note103 committed:
ユーザーコントローラーを更新

indexとshowアクションのユーザー取得機能を追加

note103 committed:
ユーザー一覧と詳細ビューを更新

ユーザー情報を表示するように変更

note103 committed:
ユーザー詳細ビューの翻訳キーを修正

日時情報が翻訳されるように変更

note103 committed:
ユーザー登録とユーザー編集のカスタムビューを作成

note103 committed:
ユーザーに郵便番号、住所、自己紹介のフィールドを追加

note103 committed:
ユーザー登録フォームに郵便番号、住所、自己紹介フィールドを追加し、翻訳キーを適用

note103 committed:
ユーザー属性に住所、郵便番号、自己紹介の翻訳キーを追加

note103 committed:
翻訳キーを使用してログイン、サインアップ、パスワードリセットリンクを更新

note103 committed:
devise-i18nのバージョンを1.12.1に固定

- I18n::InvalidPluralizationDataエラー回避のため
- 1.13.0から1.14.0ではユーザー情報更新時に確認用パスワードを入れずに更新すると、バリデーションエラーを起こさずI18n::InvalidPluralizationDataエラーになってしまう

note103 committed:
ダミーユーザーを作成するためのシードファイルを追加

note103 committed:
letter_opener_webを追加し、開発環境でのメール送信設定を更新

note103 committed:
ページネーションの設定を追加

- ユーザー一覧ビューのHTMLを修正
- ページ辺りの件数をUserモデルに設定

note103 committed:
ユーザー詳細ページに郵便番号、住所、自己紹介を表示するように更新

note103 committed:
erb_lintのバージョンを0.5.0以上に更新

- Gemfile.lock に記録されていた erb_lint 0.4.0 では `Gem::Exception` エラーが発生した
- bundler 経由で erb_lint を実行しようとすると以下のようなエラーになったため：
    bundler: failed to load command: erb_lint
    can't find executable erb_lint for gem erb_lint

note103 committed:
erb_lintの指摘に従って改行を修正

note103 committed:
Rubocopの指摘に従ってクオートのスタイルとマジックコメントを修正

note103 committed:
Deviseの導入時に生成された未使用ファイルを整理のため削除

note103 committed:
ユーザー登録およびアカウント更新時にカスタムフィールドの登録・更新を許可するパラメータを追加

note103 committed:
ユーザー登録コントローラーを追加し、ユーザー情報更新後のリダイレクト先を変更

note103 committed:
未使用のコメントアウトされたコードを削除

note103 committed:
通知メッセージの重複を削除

note103 committed:
アラートおよび通知メッセージのスタイルを追加

note103 committed:
ユーザー登録およびユーザー詳細ページの下部メニューを改善

note103 committed:
アカウント削除ボタンを移動

本の削除と構成を揃える意図

note103 committed:
ダミーユーザーのパスワードを修正

元の典型的な内容ではブラウザーから警告が出るため

note103 committed:
ナビゲーションメニューのリンクを整理

- 他ページの用語に揃えてサインアウト→ログアウトに変更
- トップページへの移動リンクを作成し、同義の本の一覧を一時的にカット
  - 今後トップ専用のページを作ったときは復活させる

note103 committed:
ナビゲーションメニューを再考、修正

- 「トップページ」では本の一覧であることがわかりづらいため戻し
- リンク間に区切りを追加
- 並び順も再考

note103 committed:
ダミーユーザー作成ロジックをseeds.rbに統合し、dummy_users.rbを削除

note103 committed:
ユーザー削除ボタンの表示ロジックを修正

どのユーザー詳細画面からもログインユーザーを削除できてしまったので

note103 committed:
ユーザーの属性としてusernameとfull_nameフィールドを追加

note103 committed:
ユーザー属性の追加に伴い、各画面の表示を変更

- ユーザー一覧でユーザー名を表示するように変更
- ユーザー詳細ページのトップをメールアドレスからユーザー名に変更
- 日本語の翻訳ファイルにusernameとfull_nameの翻訳を追加

note103 committed:
ダミーデータの作成ロジックを更新

- Fakerで使用するメソッドを変更
- ユーザー作成時にusernameとfull_nameフィールドを追加

note103 committed:
ページネーションの表示件数を5から10に変更

note103 committed:
ユーザー詳細ページから不要な属性を削除

- 作成日時
- 更新日時

note103 committed:
ダミーユーザー作成時のメールアドレス生成ロジックを修正

note103 committed:
ユーザーのモデル名を翻訳ファイルから削除

devise-i18nの同翻訳キーが使われるため

maedana commented on Gemfile:
```diff
@@ -65,9 +65,10 @@ group :development do
   # gem "spring"
 
-  gem 'erb_lint', require: false
+  gem 'erb_lint', '>= 0.5.0', require: false
```
このgemに限らずですが、特別な理由がなければバージョン指定せず最新版でよいかなーと思います。元から入っているものは今回そのままでいいので、erb_lintとdevise-i18nはバージョン指定なくしましょうー。

なお、Railsのように`bundle update`で勝手に最新版に上がってほしくないようなものはバージョン固定したほうがよいです。

note103 replied:
ありがとうございます。こちらの件、どちらも理由がなかったわけではなく、erb_lintについては以下のコミットメッセージにある経緯でした。
https://github.com/note103/fjord-books_app-2023/commit/37affbd2fb5208e1928cee1804bb5bccde595b06

Gemfile.lockのバージョンが古く、そのまま実行するとエラーが出るので `bundle update`や`bundle update erb_lint`を試したのですが、どちらも依存関係にあるgemのアップデートに伴って新たなエラーが出てしまい、困ったなと。

そこでGemfileの方でv0.5.0以上を指定して読み込んだところ、エラーを回避しつつerb_lintもアップデートできたので、ひとまずそのようにしていました。ただ、ご指摘を受けてあらためて調べたところ、 `bundle update --conservative erb_lint` をすればどちらも解消できたので、その前提で修正しました。
https://github.com/note103/fjord-books_app-2023/pull/4/commits/46be45db8eece12a5ce28f4e49974aceaab85210

---

devise-i18nの経緯については以下の日報にまとめていたのですが、
https://bootcamp.fjord.jp/reports/114775#i18n%3A%3Ainvalidpluralizationdata%E3%82%A8%E3%83%A9%E3%83%BC%E3%81%AE%E6%A0%B9%E6%9C%AC%E5%8E%9F%E5%9B%A0

結論的には、最新バージョンを使って手元に上書き用の記述を追記するか、手元の翻訳ファイルは最小限にしてdevise-i18nのバージョンを古い方に落とす（なるべくdevise-i18n内の翻訳に任せる）かを検討して、後者の方にしていました。

今回のケースでは最新版の [`ja.yml`](https://github.com/devise-i18n/devise-i18n/blob/master/rails/locales/ja.yml#L30) の方が変ではという印象があり、それを生かすために本来不要な記述を手元のファイルに追記するのはどうなのか・・という判断だったのですが、どうでしょうか。対応自体は今回のようにしつつ、以下の伊藤さんの記述のようにコメントとして残しておく、という方法もあるのかなと後から思ったのですが。
https://github.com/note103/fjord-books_app-2023/commit/f2c48a9ea0256af8c25222bccd64ece952c2f7e8#diff-d09ea66f8227784ff4393d88a19836f321c915ae10031d16c93d67e6283ab55fR19-R69

もしバージョンを未指定にすることの方を優先した方が良さそうでしたら、翻訳ファイルへの追記と合わせてあらためて修正します。

maedana replied:
なるほど、明確なご事情があったのであればOKです。ただ、その場合は
> 以下の伊藤さんの記述のようにコメントとして残しておく

のようにGemfileにコメントとして何故バージョン固定しているのか、を書いたほうが他の誰かであったり将来の自分がコードを触るときの助けとなるかもしれません。

> それを生かすために本来不要な記述を手元のファイルに追記するのはどうなのか

変なキーだけを上書きするという手段も取れるかな、、と思います。個人的には明らかにおかしいのであればgemにissueを立てるなりPRを送るなりして改善してもらって最新版を使う、という方法をオススメしたいです。(慣れてないと中々ハードルが高いかもしれませんが)

note103 replied:
ご確認ありがとうございます。今回はトラブルを解消することで手一杯になってしまい、レビューで見ていただくことや、将来的に共同で開発していくことまで頭が回っていませんでした。標準から外れたことをしている場合は、他の人が見ても納得できるようなメッセージを示すようにしたいと思います。

IssueやPRの件、たしかにですね。じつは頭の隅で「これってPRチャンスでは・・」と思ったりもしたのですが、ちょっとハードルが高いかなとも思っていました。とはいえ、いずれは貢献したい気持ちもありますので、前向きに考えたいと思っています。

maedana commented on app/views/devise/registrations/_form.html.erb:
```diff
@@ -0,0 +1,29 @@
+  <div class="field">
+    <%= f.label :email %><br />
+    <%= f.email_field :email, autofocus: true, autocomplete: "email" %>
```
ここに限らずですけども、autofocus, autocompleteについて意味を理解してつけていますか?

note103 replied:
ああ、このあたりはdevise:viewsで生成されたフォームをそのまま使用しているつもりだったので、精査していませんでした。

emailのコードをそのまま他のフィールドでもコピーして使ってしまっているので、おかしなことになっていますね。修正します。

maedana replied:
ユーザビリティにも関わる部分ですのでコピペする際は十分気をつけるようにしましょうー

maedana commented on app/views/users/show.html.erb:
```diff
@@ -0,0 +1,13 @@
+<h1><%= @user.username %></h1>
+<p><%= t('activerecord.attributes.user.full_name') %>: <%= @user.full_name %></p>
```
i18n課題でやったのと、このPR内の他の箇所でもやってますが、Railsにはmodelのカラムの国際化のために便利なメソッドが用意されてます。それを使いましょうー

note103 replied:
なるほど。tヘルパーを使わずに、ということでしょうか。思いついていませんでした。やってみます。

maedana commented on config/initializers/devise.rb:
```diff
@@ -0,0 +1,313 @@
+  # Configure the e-mail address which will be shown in Devise::Mailer,
+  # note that it will be overwritten if you use your own mailer class
+  # with default "from" parameter.
+  config.mailer_sender = 'please-change-me-at-config-initializers-devise@example.com'
```
本来はこういうところの変更も必要です。今回はいいのでご参考までに。

maedana commented on config/initializers/devise.rb:
```diff
@@ -0,0 +1,313 @@
+
+  # ==> Configuration for :validatable
+  # Range for password length.
+  config.password_length = 6..128
```
ここもデフォルトだと短すぎるんですよね。今回はこのままでいいです。

note103 replied:
ありがとうございます。気になっていたのですが、今回の対応範囲で入れた方が良いのか、ちょっと迷ってそのままにしていました。留意しておきます。

maedana commented on config/locales/devise.en.yml:
```diff
@@ -0,0 +1,65 @@
+      not_locked: "was not locked"
+      not_saved:
+        one: "1 error prohibited this %{resource} from being saved:"
+        other: "%{count} errors prohibited this %{resource} from being saved:"
```
なんのために追加しているファイルですか? 特に改変してないのであれば不要なのではないかなと。devise-i18nには各言語用ファイルが同梱されてますよー。https://github.com/tigrish/devise-i18n/tree/master/rails/locales

note103 replied:
なるほど、deviseのインストール時に自動生成されて、中身も記載されていたので必要なものだと思い込んでいました。たしかに不要ですね。外しておきます。

maedana commented on config/routes.rb:
```diff
@@ -1,7 +1,10 @@
+  }
   resources :books
-  # Define your application routes per the DSL in https://guides.rubyonrails.org/routing.html
+  resources :users, only: [:index, :show]
```
ちゃんと最低限のroutingに絞ってますね :+1:

note103 replied:
ありがとうございます！

maedana commented on db/migrate/20250713154553_add_column_to_user.rb:
```diff
@@ -0,0 +1,6 @@
+class AddColumnToUser < ActiveRecord::Migration[7.0]
```
細かいですが、どんなパターンのカラム追加でもよさそうな命名になってますね。migrationファイルはもっと具体的でユニークな命名にしたほうが後々困らない気がします。今回はこのままでよいですが。

note103 replied:
ありがとうございます。その部分、マイグレーションファイルの生成時に便宜的につける名称という程度の認識でした。クラス名になることを考えたら、汎用性は低くした方が良いですね。留意しておきます。

maedana replied:
長期で運用するプロジェクトになってくるとmigrationファイルだけで100を超えるとかあったりしますし、同じテーブルに対する変更も複数あるのが当たり前になってきますので、今後意識していけるとよいと思います :smile:

note103 replied:
ありがとうございます。マイグレーション、まだ理解が不十分な面もあり、イメージしきれていませんでした。そのあたり、意識してみます。

note103 committed:
ユーザー詳細ページの属性名の翻訳を変更

tヘルパーを使ったものからModel.human_attribute_nameを用いたものへ

note103 committed:
フォームのフィールドにおいてautofocus, autocompleteの設定を修正

note103 committed:
Gemfileからerb_lintのバージョン指定を削除してerb_lintのみbundle update

- `bundle update --conservative erb_lint`
- `bundle update erb_lint` を行うとRailsコンソールを使えなくなるため

note103 committed:
不要なdevise.en.ymlを削除

maedana commented on app/views/devise/registrations/_form.html.erb:
```diff
@@ -5,25 +5,25 @@
   <div class="field">
     <%= f.label :address %><br />
-    <%= f.text_field :address, autofocus: true, autocomplete: t('attributes.user.address') %>
+    <%= f.text_field :address, autocomplete: "street-address" %>
```
autocomplete名とフィールド名が一致していたりそうでなかったりで一貫性がないように見えるのが少し気になりました。

maedana commented on app/views/users/show.html.erb:
```diff
@@ -1,9 +1,9 @@
-<p><%= t('activerecord.attributes.user.postal_code') %>: <%= @user.postal_code %></p>
-<p><%= t('activerecord.attributes.user.address') %>: <%= @user.address %></p>
-<p><%= t('activerecord.attributes.user.bio') %>: <%= @user.bio %></p>
+<p><%= User.human_attribute_name('full_name') %>: <%= @user.full_name %></p>
```
細かいですが、シンボルでかけるときはシンボルで書きたいですね。この場合 `:full_name`でいいですね。他も同様です。
